doc_src/prompt.rst

```diff
@@ -0,0 +1,189 @@
+
+Our first prompt
+----------------
+
+Let's look at a very simple example::
+
+  function fish_prompt
+      echo $PWD '>'
+  end
+
+This just prints the current working directory (:envvar:`PWD`) and a ``>`` symbol to show where the prompt ends. The ``>`` is :ref:`quoted <quotes>` because otherwise it would signify a :ref:`redirection <redirects>`.
+
+Because we've use :doc:`echo <cmds/echo>`, it adds spaces between the two so it ends up looking like (assuming ``_`` is your cursor):
+
+.. role:: white
+.. parsed-literal::
+    :class: highlight
+
+    :white:`/home/tutorial >`\ _
+
+Formatting
+----------
+
+If you don't want the space, there are other ways to print this:
+
+- ``echo -s``, which tells echo to not print a space.
+- We could construct one string: ``echo "$PWD>"``
+- :doc:`printf <cmds/printf>`, which allows us to give a format string to format our... string
+- :doc:`string join <cmds/string-join>`, if we want to have one thing appear between all our components.
+
+For now, let's use ``string join ''`` to write all our components without anything inbetween::
+
+  function fish_prompt
+      string join '' -- $PWD '>'
+  end
+
+The ``--`` tells ``string`` that no options can come after it, in case we extend this with something that can start with a ``-``.
+
+Adding color
+------------
+
+This prompt is functional, but a bit boring. We could add some color.
+
+In general, telling a terminal to print text in color involves sending it an escape sequence. For example you could send it
+
+.. role:: red
+.. parsed-literal::
+    :class: highlight
+
+    ESCAPE [ 3 1 m
+
+where "escape" is the ascii escape character stored as 0x1b.
+
+We could do this in fish, by writing escape as ``\e`` or ``\x1b``::
+
+  echo \e\[31mfoo
+
+This will tell the terminal to print "foo" in red and will look like this:
+
+.. parsed-literal::
+    :class: highlight
+
+    :red:`foo`
+
+To make it blue, we would use ``\e\[34mfoo``. However, obviously it's a bit annoying to have to remember the numbers for these colors, and modern terminals can do quite a bit more - most can not only do 256 colors but even 24-bit RGB.
+
+So, fish offers the :doc:`set_color <cmds/set_color>` command instead, so you can do::
+
+  echo (set_color red)foo
+
+``set_color`` can also handle RGB colors like ``set_color 23b455``, and other formatting options including bold and italics.
+
+So, taking our previous prompt and adding some color::
+
+  function fish_prompt
+      string join '' -- (set_color green) $PWD (set_color normal) '>'
+  end
+
+A "normal" color tells the terminal to go back to its normal formatting options.
+
+Shortening $PWD
+---------------
+
+This is alright, but our $PWD can be a bit long, and we are typically only interested in the last few directories. We can shorten this with the :doc:`prompt_pwd <cmds/prompt_pwd>` helper that will give us a shortened working directory::
```

Comment:
Alright's status as a word is controversial - what about "fine"? I'd use the envvar link here for PWD.